### PR TITLE
Split insert block function

### LINKED ISSUE #9959: split `DatabaseProvider::insert_block` to isolate transaction-specific writing
In the new persistence / DB model we will not be first writing transactions to DB and gradually moving to static files. Instead, we will be directly writing transaction data to static files.

The `insert_block` method by default writes to the transaction and headers tables:
https://github.com/paradigmxyz/reth/blob/02d25304f9c3352474a7f57e10467ae7a4038c1f/crates/storage/provider/src/providers/database/provider.rs#L3431

If we split the method, with the first being for transaction + header data, and the other method for everything else, we can use the second method instead of `insert_block` in the persistence task.

This may also be of use for the bodies stage:
https://github.com/paradigmxyz/reth/blob/0fece98b0529788b56b24e806046d585b0517ea9/crates/stages/stages/src/stages/bodies.rs#L115

### PR DESCRIPTION
**Pull request by elijahhampton**
Creates two functions "insert_block_additional_data" and "insert_block_header_and_transaction_data" to split insert_block.

Fixes https://github.com/paradigmxyz/reth/issues/9959

### CODE REVIEW

**Review by mattsse — COMMENTED** [review]
imo this makes sense, wdyt @Rjected 

could you please undo the additional ;?

File: crates/storage/provider/src/providers/database/provider.rs [hunk 1]
```diff
@@ -3550,6 +3548,49 @@ impl<TX: DbTxMut + DbTx> BlockWriter for DatabaseProvider<TX> {
...
+    ///
+    /// # Arguments
+    /// * `block` - Reference to the [`SealedBlockWithSenders`] containing the additional data
+    ///
+    /// # Returns
+    /// * `ProviderResult<()>` - Ok if the insertion was successful, or an error if it failed
+    fn insert_block_additional_data(
+        &self,
+        block: SealedBlockWithSenders,
+        durations_recorder: &mut metrics::DurationsRecorder,
```
Comment (Rjected) [inline crates/storage/provider/src/providers/database/provider.rs:3581]: let's use the full import at the top:
```rust
import metrics::DurationsRecorder
```

File: crates/storage/provider/src/providers/database/provider.rs [hunk 2]
```diff
@@ -3567,14 +3608,40 @@ impl<TX: DbTxMut + DbTx> BlockWriter for DatabaseProvider<TX> {
...
+    fn insert_block(
+        &self,
+        block: SealedBlockWithSenders,
+    ) -> ProviderResult<StoredBlockBodyIndices> {
+        let block_number = block.number;
+        let mut durations_recorder = metrics::DurationsRecorder::default();
+
+        let block_indices =
+            self.insert_block_header_and_transaction_data(block.clone(), &mut durations_recorder)?;
+        self.insert_block_additional_data(block, &mut durations_recorder)?;
```
Comment (Rjected) [inline crates/storage/provider/src/providers/database/provider.rs:3644]: do we need to clone here? or can we make `insert_block_header_and_transaction_data` accept `&SealedBlockWithSenders`

**Review by Rjected — CHANGES_REQUESTED** [review]
awesome! I just have a few questions

File: crates/storage/provider/src/traits/block.rs [hunk 3]
```diff
@@ -50,6 +52,58 @@ pub trait BlockWriter: Send + Sync {
...
+    /// - `ProviderResult<()>` - Ok if the insertion was successful, or an error if it failed
+    ///
+    /// # Tables modified
+    /// - `BlockOmmers` - If the block contains ommers
+    /// - `BlockWithdrawals` - If the block contains withdrawals
+    /// - `BlockRequests` - If the block contains requests
+    fn insert_block_additional_data(
+        &self,
+        block: SealedBlockWithSenders,
+        durations_recorder: &mut metrics::DurationsRecorder,
```
Comment (Rjected) [inline crates/storage/provider/src/traits/block.rs:74]: I also don't think this needs to accept a `&mut metrics::DurationsRecorder`, we can just initialize it as `DurationsRecorder::default()` in each method's implementation, similar to how we do it now in `insert_block`

Comment (Rjected) [reply crates/storage/provider/src/traits/block.rs:74]: actually nevermind, this interface is correct

**Comment by elijahhampton** [comment]
@mattsse Are there other changes you would like to see for this PR?

**Review by Rjected — APPROVED** [review]
@elijahhampton sorry for not getting back to the review in a timely manner, this looks good to me! Should just need a rebase to solve merge conflicts

**Comment by elijahhampton** [comment]
> @elijahhampton sorry for not getting back to the review in a timely manner, this looks good to me! Should just need a rebase to solve merge conflicts

Sweet. Good to go.

**Comment by Rjected** [comment]
just trying to make sure we can merge this - I accidentally re-introduced the methods that were removed from https://github.com/paradigmxyz/reth/pull/12367, so we just need to use the traits for those again

**Comment by Rjected** [comment]
sorry for not getting to this in time, this was my fault :( 

closing this because I think https://github.com/paradigmxyz/reth/issues/9959 is actually solved by https://github.com/paradigmxyz/reth/pull/12694 and its integration

**Pull request closed without merging**